---
title: "Systemix MAG-SMAG Profiling"
author: "Moises A. Rojas"
date: "2023-12-01 // mod. Mar 2024 // rev. Jun 2024"
output: github_document
---

```{r setup}
knitr::opts_knit$set(root.dir = "/Users/moises/Downloads/")

suppressMessages(library(ape)) #tree basics
suppressMessages(library(treeio)) #join trees
suppressMessages(library(ggtree)) #tree plotting
suppressMessages(library(ggplot2)) #plots
suppressMessages(library(ggtreeExtra)) #geom_fruit
suppressMessages(library(ggnewscale)) #scales in ggplot2
suppressMessages(library(ggstar)) #geom_star for ggplot2
suppressMessages(library(tidyverse)) #strings, dataframes processing
suppressMessages(library(magrittr)) #pipes %>%
suppressMessages(library(tidytree)) #trees as dataframes
suppressMessages(library(RColorBrewer)) #color palettes
suppressWarnings(library(colorspace)) #colors manipulation
suppressWarnings(library(patchwork)) #side-by-side plotting
suppressWarnings(library(openxlsx)) #read excel
suppressWarnings(library(ComplexHeatmap)) #heatmap
suppressWarnings(library(PCAtools)) #pca
suppressWarnings(library(ggbiplot)) #pca plots
suppressWarnings(library(agricolae)) #fisher's test
suppressWarnings(library(ggpubr)) #stats & plots
suppressWarnings(library(dndscv)) #dN/dS of snps
suppressWarnings(library(clusterProfiler)) #gene enrichment
suppressWarnings(library(circlize)) #circular viz
```

## Prepare and merge MAG-SMAG metadata

```{r magdata}
# Read Systemix MAGs taxonomy and QC files
gtdb_taxonomy <- read.csv("article_profiling/1-taxonomy/gtdbtk_classification.tsv", sep = "\t", stringsAsFactors = FALSE)
mags_qc <- read.csv("article_profiling/2-statistics/checkm_parsed_bins.tsv", sep = "\t")

# Inspect completeness and contamination per sample, then select one
filter(mags_qc, grepl("\\-S15\\.", bin)) %>% dplyr::select(2:3) %>% summary()
mag_samples <- filter(gtdb_taxonomy, grepl("\\-S15\\.", genome))

# Add QC attributes
colnames(mag_samples)[1] <- "bin"
mag_samples <- left_join(mag_samples, mags_qc, by = c("bin"))

# Parse columns
mag_samples <- dplyr::select(mag_samples, -c("taxonomy"))
mag_samples <- mag_samples %>%
  mutate_if(is.character, function(x) gsub("d__|p__|c__|o__|f__|g__|s__", "", x))
mag_samples <- mag_samples %>% separate(bin, "\\.", into = c("mag_name", "mag_id"))

# Define final attributes
mag_samples["mag_num"] <- nrow(mag_samples)
mag_samples["ecosystem"] <- "Biomining"
mag_samples["continent"] <- "South America"

# Load SMAG catalog (metadata from tables 1-2)
smag_md1 <- read.xlsx("/mnt/beegfs/home/mrojas/DiGenomaLab/Systemix/SMAG/Extended_Data_Table_1.xlsx")
smag_md2 <- read.xlsx("/mnt/beegfs/home/mrojas/DiGenomaLab/Systemix/SMAG/Extended_Data_Table_2.xlsx")

# Save attributes
smag_md1_sel <- smag_md1 %>% dplyr::select("mag_name", "mag_num", "Ecosystem", "Continent")
smag_md2_sel <- smag_md2 %>%
  dplyr::select("user_genome", "completeness", "contamination",
                "strain_heterogeneity", "length", "N50", "classification",
                "phylum", "class", "order", "family", "genus", "species")

# Merge selections and parse taxa
smag_data <- smag_md2_sel %>% separate(user_genome, "\\.", into = c("mag_name", "mag_id"))
smag_data <- left_join(smag_data, smag_md1_sel, by = c("mag_name"))
smag_data <- smag_data %>%
  mutate_if(is.character, function(x) gsub("d__|p__|c__|o__|f__|g__|s__", "", x))

# Rename columns
colnames(smag_data)[6] <- "genome_size"
colnames(smag_data)[8] <- "kingdom"
colnames(smag_data)[16] <- "ecosystem"
colnames(smag_data)[17] <- "continent"
colnames(smag_data)[7] <- "kingdom"
colnames(smag_data)[14] <- "ecosystem"
colnames(smag_data)[15] <- "continent"

# Merge metadata and save to file
mags_metadata <- merge(mag_samples, smag_data, all = TRUE)
if (file.exists("4-mags-smag/mags_metadata.tsv") == FALSE) {
  write.table(mags_metadata, "4-mags-smag/mags_metadata.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
}

# Save parsed SMAG metadata
if (file.exists("4-mags-smag/smag_data.tsv") == FALSE) {
  write.table(smag_data, "4-mags-smag/smag_data.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
}

# Save biomining metadata (S15)
if (file.exists("4-mags-smag/mags_biomining.tsv") == FALSE) {
  write.table(mag_samples, "4-mags-smag/mags_biomining.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
}

# Clean memory
rm(smag_md1, smag_md2, smag_md1_sel, smag_md2_sel)
#rm(list = ls(pattern = "^kegg_hm_cols_"))
```

## Section 1: Taxonomic analysis 

### In-depth biomining MAGs vs SMAG comparison

```{r magcombtaxa}
# Biomining provides a new phylum -> Nitrospirota_A
mag_samples_new_phy <- mag_samples[!(mag_samples$phylum %in% smag_data$phylum), ]

# Explore metadata: some bins are not assigned to ecosystems
mags_metadata %>%
  dplyr::filter(is.na(ecosystem)) %>%
  dplyr::group_by(mag_name) %>%
  dplyr::summarise() #n=401

# Keep only bacteria and known ecosystems (n = 32975)
mags_metadata_bac <- mags_metadata %>% dplyr::filter(kingdom == "Bacteria" & !is.na(ecosystem))

# Save bacterial metadata to file
if (file.exists("4-mags-smag/mags_metadata_bac.tsv") == FALSE) {
  write.table(mags_metadata_bac, "4-mags-smag/mags_metadata_bac.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
}

# Select top phyla and plot a histogram
mags_md_top_phy <- table(mags_metadata_bac$phylum) %>%
  sort(TRUE) %>%
  head(n = 40) %>%
  as.data.frame() %>%
  dplyr::rename(phy_top = Var1, freq = Freq)

plot_depth_top_phyla <- mags_md_top_phy %>%
  ggplot(aes(x = freq, y = reorder(phy_top, freq))) +
  geom_bar(stat = "identity", fill = "#459682") +
  geom_text(aes(label = freq), vjust = 0.5, hjust = -0.25, size = 3) +
  labs(x = "Number", y = "Phylum") +
  scale_x_continuous(limits = c(0, 10000)) +
  theme_classic() +
  theme(axis.text = element_text(size = 14),
        axis.title = element_text(size = 16))
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_top_phyla.pdf", plot_depth_top_phyla, height = 7, width = 6)

# Filter metadata..
# Set colors for ecosystem
#mags_colors_eco <- colorRampPalette(brewer.pal(12, "Paired"))(length(unique(mags_metadata_bac$ecosystem)))
#mags_colors_eco <- sample(mags_colors_eco)
mags_colors_eco <- c("#1F78B4", "#FF7F00", "#B2DF8A", "#E31A1C", "#A6CEE3",
                     "#6A3D9A", "#33A02C", "#CAB2D6", "#FDBF6F", "#FB9A99")

# Which bacteria from biomining sites are annotated as known?
mags_depth_spec_biom <- mags_metadata_bac %>%
  dplyr::select(mag_name, family, genus, species, ecosystem) %>%
  dplyr::filter(species != "" & ecosystem == "Biomining")

#are they present in other SMAG?
mags_depth_spec <- mags_metadata_bac[mags_metadata_bac$species %in%
                                     mags_depth_spec_biom$species, ] %>%
  dplyr::select(mag_name, mag_id, phylum, family, genus, species, ecosystem, continent) %>%
  dplyr::group_by(ecosystem)

plot_depth_spec <- ggplot(mags_depth_spec) +
  geom_bar(aes(x = species, fill = phylum)) +
  labs(x = "", y = "Number of kSGBs from biomining") +
  scale_fill_manual(values = "#78acc2") +
  coord_flip() +
  theme_classic() +
  theme(legend.title = element_blank(),
        legend.position = c(0.88, 0.12),
        strip.background = element_blank(),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14),
        strip.text.x = element_text(size = 10)) +
  facet_grid(.~ecosystem)
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_biom_species.pdf", plot_depth_spec, height = 3, width = 9)

# Which ecosystem has more known species (# and ratio)?
mags_depth_spec_eco <- mags_metadata_bac %>%
  dplyr::filter(species != "") %>%
  dplyr::group_by(species, ecosystem) %>%
  dplyr::summarise(count_sp = table(species)) %>% #.groups = "drop"
  ungroup() %>%
  dplyr::group_by(ecosystem) %>%
  dplyr::summarise(count_eco = sum(count_sp)) %>%
  dplyr::arrange(desc(count_eco)) #max 1821

#count number of MAGs per ecosystem
mags_depth_spec_eco_n <- mags_metadata_bac %>%
  group_by(ecosystem) %>%
  dplyr::summarise(mags = n_distinct(mag_name))
mags_depth_spec_eco <- left_join(mags_depth_spec_eco, mags_depth_spec_eco_n, by = c("ecosystem"))
rm(mags_depth_spec_eco_n) #clean memory

#add column
mags_depth_spec_eco <- mags_depth_spec_eco %>%
  dplyr::mutate(mags = replace(mags, ecosystem == "Biomining", 1)) %>%
  dplyr::mutate(ratio = count_eco / mags) #max 6.52

plot_depth_spec_eco <- mags_depth_spec_eco %>%
  ggplot(aes(x = reorder(ecosystem, ratio), y = ratio, fill = ecosystem)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = round(ratio, 2)), vjust = 0.5, hjust = -0.25, size = 3) +
  labs(x = "Ecosystem", y = "Number of kSGBs/MAGs") +
  scale_y_continuous(limits = c(0, 10)) +
  scale_fill_manual(values = mags_colors_eco) +
  coord_flip() +
  theme_classic() +
  theme(legend.position = "none",
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14))
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_species_eco.pdf", plot_depth_spec_eco, height = 7, width = 6)

# What about unknown/known species-level genome bins (SGBs) at phylum level?
mags_depth_sgbs <- mags_metadata_bac %>%
  dplyr::group_by(phylum, species) %>%
  dplyr::summarise(n_spe = n()) %>%
  ungroup() %>%
  dplyr::group_by(phylum) %>%
  dplyr::mutate(t_phy = sum(n_spe)) %>% ###
  dplyr::filter(species == "") %>%
  dplyr::mutate(diff_k = t_phy - n_spe) %>%
  dplyr::mutate(perc_u = n_spe * 100 / t_phy) %>%
  dplyr::arrange(desc(t_phy)) %>%
  pivot_longer(cols = c(n_spe, diff_k), names_to = "cond", values_to = "tot")

plot_depth_sgbs <- mags_depth_sgbs[1:60, ] %>%
  ggplot(aes(x = reorder(phylum, tot), y = tot, fill = cond)) +
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_fill_manual(name = "", labels = c("kSGB", "uSGB"), values = c("#dd7e8b", "#459682")) +
  scale_y_continuous(limits = c(0, 8000)) +
  labs(x = "Phylum", y = "Number of species") +
  coord_flip() +
  theme_classic() +
  theme(legend.position = c(0.85, 0.15),
        strip.placement = "outside",
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14))

#add % column
mags_depth_sgbs_p <- mags_depth_sgbs %>%
  dplyr::select(-c("species")) %>%
  dplyr::mutate(perc_k = 100 - perc_u) %>%
  pivot_longer(cols = c(perc_u, perc_k), names_to = "cond_p", values_to = "tot_p") %>%
  dplyr::filter(cond == "n_spe") #just to remove 1 row per phylum/cond

plot_depth_sgbs_p <- mags_depth_sgbs_p[1:60, ] %>%
  ggplot(aes(x = reorder(phylum, t_phy), y = tot_p, fill = cond_p)) +
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_fill_manual(name = "", labels = c("kSGB", "uSGB"), values = c("#dd7e8b", "#459682")) +
  scale_y_continuous(limits = c(0, 100)) +
  labs(x = "", y = "% of species") +
  coord_flip() +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.y = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14))

#merge
plot_depth_sgbs_both <- wrap_plots(plot_depth_sgbs, plot_depth_sgbs_p,
                                   nrow = 1, ncol = 2, widths = c(0.55, 0.5))
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_SGBs_phylum.pdf", plot_depth_sgbs_both, height = 7, width = 7)

# And unknown/known SGBs for biomining?
mags_depth_sgbs_biom <- mags_metadata_bac %>%
  dplyr::filter(ecosystem == "Biomining") %>%
  dplyr::group_by(phylum, species) %>%
  dplyr::summarise(n_spe = n()) %>%
  ungroup() %>%
  dplyr::group_by(phylum) %>%
  dplyr::mutate(t_phy = sum(n_spe)) %>%
  dplyr::filter(species == "") %>%
  dplyr::mutate(diff_k = t_phy - n_spe) %>%
  dplyr::mutate(perc_u = n_spe * 100 / t_phy) %>%
  dplyr::arrange(desc(t_phy)) %>%
  pivot_longer(cols = c(n_spe, diff_k), names_to = "cond", values_to = "tot")

plot_depth_sgbs_biom <- mags_depth_sgbs_biom %>%
  ggplot(aes(x = reorder(phylum, tot), y = tot, fill = cond)) +
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_fill_manual(name = "", labels = c("kSGB", "uSGB"), values = c("#dd7e8b", "#459682")) +
  scale_y_continuous(limits = c(0, 25)) +
  labs(x = "Phylum", y = "Biomining species") +
  coord_flip() +
  theme_classic() +
  theme(aspect.ratio = 0.875,
        legend.position = c(0.75, 0.3),
        strip.placement = "outside",
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14))
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_SGBs_phylum_biom.pdf", plot_depth_sgbs_biom, height = 4, width = 6)

# What is the distribution of uSGBs by ecosystem?
mags_depth_usgbs_eco <- mags_metadata_bac %>%
  dplyr::filter(species == "") %>%
  dplyr::group_by(phylum, ecosystem) %>%
  dplyr::summarise(num = table(ecosystem)) %>%
  dplyr::mutate(phylum = replace(phylum, num < 5, "Other")) #group lows as "other" -> summarise?

plot_depth_usgbs <- mags_depth_usgbs_eco %>%
  ggplot(aes(x = num, y = reorder(phylum, num), fill = ecosystem)) +
  geom_col() +
  #geom_text(aes(label = num), vjust = 0.5, hjust = -0.5, size = 2.5) +
  labs(x = "Number of uSGBs", y = "Phylum") +
  scale_x_continuous(n.breaks = 3) +
  scale_fill_manual(values = mags_colors_eco) +
  theme_classic() +
  theme(legend.position = "none",
        strip.placement = "outside",
        strip.background = element_blank(),
        panel.border = element_blank(),
        axis.text = element_text(size = 14),
        axis.title = element_text(size = 16),
        strip.text.x = element_text(size = 10)) +
  facet_grid(.~ecosystem, scales = "free")
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_uSGBs_eco.pdf", plot_depth_usgbs, height = 9, width = 14)

# What about the annotated bins at genus level (kGGBs) from biomining?
mags_depth_kggbs_biom <- mags_metadata_bac %>%
  dplyr::select(mag_name, mag_id, phylum, family, genus, ecosystem) %>%
  dplyr::filter(genus != "" & ecosystem == "Biomining") #nrow=21

#match with the other ecosystems
mags_depth_kggbs_biom <- mags_metadata_bac[mags_metadata_bac$genus %in%
                                           mags_depth_kggbs_biom$genus, ] %>%
  dplyr::select(mag_name, mag_id, phylum, family, genus, ecosystem) %>%
  dplyr::group_by(genus) %>%
  dplyr::mutate(tot_gen = n()) %>%
  dplyr::arrange(desc(tot_gen)) #nrow=535

#scale to 0-100%
mags_depth_kggbs_biom <- mags_depth_kggbs_biom %>%
  dplyr::group_by(genus, ecosystem) %>%
  dplyr::mutate(sum = n()) %>% #sum genus per ecosystem
  dplyr::mutate(norm = sum / tot_gen * 100) %>%
  distinct(genus, ecosystem, .keep_all = TRUE) #reduce to uniques (plot scale!)

plot_depth_kggbs_biom <- mags_depth_kggbs_biom %>%
  ggplot(aes(y = reorder(genus, norm), x = norm, fill = ecosystem)) +
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_x_continuous(limits = c(0, 101), n.breaks = 5) +
  scale_fill_manual(values = mags_colors_eco) +
  labs(x = "% kGGBs", y = "Known genus from biomining") +
  theme_classic() +
  theme(legend.title = element_blank(),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14))
ggsave("article_profiling/4-mags-smag/1-plots-taxonomy/plot_md_kGGBs_biom.pdf", plot_depth_kggbs_biom, height = 7, width = 6)

### --EXPLORE OTHER TAXA-- ###
##inspect unannotated family-level bins (uFGBs)? #1166
##what about novel bacteria at genus level?
#mags_depth_novel_gen <- mags_metadata_bac %>%
#  dplyr::filter(genus == "") %>%
#  dplyr::group_by(phylum, genus, ecosystem) %>%
#  dplyr::summarise(gen_unk = n()) %>%
#  #dplyr::arrange(desc(gen)) %>%
#  ungroup() %>%
#  dplyr::group_by(genus) %>%
#  dplyr::mutate(tot_gen = sum(gen_unk)) %>%
#  dplyr::mutate(perc = gen_unk * 100 / tot_gen)
#
##get unique classes of unknown genus by ecosystem
##Leptospirillia, Sulfobacillia
#mags_md_unk_gen2 <- mags_metadata_bac %>%
#  dplyr::filter(genus == "") %>%
#  dplyr::group_by(class, ecosystem) %>%
#  tally() %>%
#  #slice(n()) %>%
#  ungroup()
#unique(mags_md_unk_gen2[mags_md_unk_gen2[, "ecosystem"] == "Biomining", "class"])
#
##unique genus from biomining: Acidithrix, Ferrithrix, SCQM01
#unique(mag_samples[mag_samples[, "family"] != "", "genus"])
#mags_metadata_bac %>% filter(grepl("Acidithrix", genus, ignore.case = TRUE) == TRUE)
#
#plot_md_unk_cla <- ggplot(mags_md_unk_gen2[mags_md_unk_gen2$n > 10, ],
#                          aes(x = reorder(class, n), y = n, fill = ecosystem)) +
#  geom_bar(stat = "identity") +
#  labs(x = "Class", y = "Count") +
#  scale_y_continuous(limits = c(0, 250)) +
#  theme_minimal() +
#  coord_flip() +
#  facet_wrap(~ecosystem)
#ggsave("plot_md_unk_class_eco.pdf", plot_md_unk_cla, height = 7, width = 10)
#
##class of interest for acidic soils:
##Alphaproteobacteria, Gammaproteobacteria
##Acidobacteriae, Thermoleophilia, Sulfobacillia, Acidimicrobiia
#table(mags_metadata_bac[mags_metadata_bac[, "phylum"] == "Proteobacteria", "class"])
#table(mags_metadata_bac[mags_metadata_bac[, "class"] == "Acidimicrobiia", "order"])
#
##family of interest reported for Systemix [ref 2022]:
##Pseudomonadaceae (mayor), Flavobacteriaceae, Erwiniaceae
#mags_md_fams_acid <- mags_metadata_bac %>%
#  dplyr::filter(family == "Pseudomonadaceae") %>%
#  dplyr::group_by(ecosystem) %>%
#  dplyr::summarise(count = n())
```

### QC

```{r magbiomqc}
# Read QC file from CheckM
mags_qc <- read.csv("article_profiling/2-statistics/checkm_parsed_bins.tsv", sep = "\t")

# Select S15 sample
mags_qc <- mags_qc %>% filter(grepl("Refined-S15", bin))

# Parse QCs
mags_qc_hq <- mags_qc %>%
  dplyr::filter(completeness >= 90 & contamination <= 5) %>%
  dplyr::summarise(mags = n(), contigs = sum(contigs),
                   avg_comp = mean(completeness), avg_cont = mean(contamination))
mags_qc_hq$label = "High Quality"
mags_qc_mq <- mags_qc %>%
  dplyr::filter(completeness < 90 & completeness >= 50 & contamination <= 10) %>%
  dplyr::summarise(mags = n(), contigs = sum(contigs),
                   avg_comp = mean(completeness), avg_cont = mean(contamination))
mags_qc_mq$label = "Medium Quality"
mags_qc_lq <- mags_qc %>%
  dplyr::filter(completeness < 50 & contamination <= 10) %>%
  dplyr::summarise(mags = n(), contigs = sum(contigs),
                   avg_comp = mean(completeness), avg_cont = mean(contamination))
mags_qc_lq$label = "Low Quality"
#merge
mags_qcs <- rbind(mags_qc_hq, mags_qc_mq, mags_qc_lq)
rm(mags_qc_hq, mags_qc_mq, mags_qc_lq) #clean memory

# Plot
plot_mags_qc <- mags_qcs %>%
  ggplot(aes(x = mags, y = fct_reorder(label, mags), fill = label)) +
  geom_col(position = "dodge2") +
  #geom_text(aes(label = mags), vjust = 0.5, hjust = -0.25, size = 3) +
  scale_fill_manual(values = c("#db7971", "#5ca650", "#859ac7")) +
  labs(x = "Number of MAGs", y = "") +
  theme_classic() +
  theme(aspect.ratio = 0.875,
        axis.text = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        legend.position = "none")
ggsave("article_profiling/2-statistics/plots/plot_biom_MAGs_QC.pdf", plot_mags_qc, height = 4, width = 6)
```

### Phylogenetic tree

```{r magbiomtree}
# Load GTDB taxonomy
gtdb_taxonomy <- read.csv("article_profiling/1-taxonomy/gtdbtk_classification.tsv", sep = "\t", stringsAsFactors = FALSE)

# Subset taxonomy to selected S15 sample and remove underscores
gtdb_taxonomy <- gtdb_taxonomy %>%
  dplyr::filter(grepl("Refined-S15", genome) == TRUE) %>%
  dplyr::select(-c("taxonomy")) %>%
  mutate_all(function(x) gsub("d__|p__|c__|o__|f__|g__|s__", "", x))

# Read tree (S15 sample), subset it to biomining placements and save to file
tree <- ape::read.tree("article_profiling/1-taxonomy/raw/gtdbtk-trees/gtdbtk.DASTool-S15.tree")
tree_s15 <- keep.tip(tree, as.character(gtdb_taxonomy$genome))
tree_s15 #44 tips
if (file.exists("article_profiling/1-taxonomy/trees-subset/gtdbtk.DASTool-S15.tree") == FALSE) {
  ape::write.tree(tree_s15, "article_profiling/1-taxonomy/trees-subset/gtdbtk.DASTool-S15.tree")
}
rm(tree) #clean memory

# Rename column from GTDB data to match with the tree
colnames(gtdb_taxonomy)[1] <- "label"

# Convert subset tree to dataframe and join some taxa
tree_s15_df <- tidytree::as_tibble(tree_s15)
tree_s15_df <- treeio::full_join(tree_s15_df, gtdb_taxonomy[c("label", "phylum", "class")], by = "label")

# Summarise phyla
tree_s15_sum_df <- tree_s15_df %>%
  filter(!is.na(phylum)) %>%
  dplyr::group_by(phylum) %>%
  summarise(n = n()) %>%
  dplyr::mutate(phylum_n = paste0(phylum, " (", n, ")"))
tree_s15_df <- left_join(tree_s15_df, tree_s15_sum_df, by = c("phylum"))
rm(tree_s15_sum_df) #clean

# Back tree to phylogenetic object
tree_s15_md <- tidytree::as.treedata(tree_s15_df)

# Assign colors to each phylum
tree_colors <- colorRampPalette(brewer.pal(12, "Paired"))(length(unique(gtdb_taxonomy$phylum)))

# Plot circular tree
#plot_tree_s15 <- ggtree(tree_s15_md, layout = "circular", size = 0.2)
# Plot tree adding phylum
#plot_tree_s15_phyla <- plot_tree_s15 +
#  geom_tippoint(aes(color = phylum), na.rm = TRUE, size = 1) +
#  geom_tiplab(aes(label = phylum, color = phylum), size = 1.75, hjust = -0.1) +
#  scale_colour_discrete(breaks = c("c", "d"), na.value = NA) +
#  scale_colour_manual(values = tree_colors, na.translate = FALSE, guide = "none") +
  #guides(color = guide_legend(title = "Phyla", ncol = 1, order = 1)) +
#  theme()
#ggsave("plot_biom_MAGs_tree_phyla.pdf", plot_tree_s15_phyla, height = 5, width = 7)

# Process QC attributes to match tree tips order
tree_tips <- tree_s15$tip.label
tree_com <- vector()
tree_con <- vector()
tree_st  <- vector()
tree_gs  <- vector()
tree_gc  <- vector()
tree_pg  <- vector()
for (i in 1:nrow(tree_s15_df)) {
  for (j in 1:nrow(mags_qc)) {
    if (tree_s15_df$label[i] == mags_qc$bin[j]) {
      tree_com <- append(tree_com, mags_qc$completeness[j])
      tree_con <- append(tree_con, mags_qc$contamination[j])
      tree_st  <- append(tree_st,  mags_qc$strain_heterogeneity[j])
      tree_gs  <- append(tree_gs,  mags_qc$genome_size[j])
      tree_gc  <- append(tree_gc,  mags_qc$gc[j])
      tree_pg  <- append(tree_pg,  mags_qc$predicted_genes[j])
    }
  }
}

# Save attributes in a dataframe
tree_qc_df <- tibble(Bin = tree_tips, Completeness = tree_com,
                     Contamination = tree_con, Strain_Het = tree_st,
                     Genome_Size = tree_gs, GC = tree_gc, Pred_Genes = tree_pg)
rm(tree_tips, tree_com, tree_con, tree_st, tree_gs, tree_gc, tree_pg, i, j) #clean memory

# Plot tree adding QC as multiple layers
plot_tree_s15_md <- ggtree(tree_s15_md, layout = "circular", size = 0.1) +
  geom_tippoint(aes(color = phylum), na.rm = TRUE, size = 1) +
  geom_tiplab(aes(label = phylum, color = phylum), size = 2, hjust = -0.2) +
  geom_treescale(x = -0.15, y = 0, offset = 0.2, fontsize = 1.4, linesize = 0.1, width = 0.1) +
  scale_colour_manual(values = tree_colors, guide = "none") +
  #add QC as layers
  new_scale_fill() + new_scale_color() +
  geom_fruit(data = tree_qc_df, geom = geom_bar,
             mapping = aes(y = Bin, x = Completeness, fill = Completeness),
             orientation = "y", stat = "identity", position = "auto",
             grid.params = list(linetype = 1, size = 0.1), pwidth = 0.3, offset = 1,
             axis.params = list(axis = "x", text.size = 1.8, vjust = 1, nbreak = 3)) +
  scale_fill_gradient(name = "Completeness (%)", low = "#1f4060", high = "#4497dd",
                      guide = guide_legend(keyheight = 0.3, keywidth = 0.6, reverse = TRUE, order = 1)) +
  new_scale_fill() + new_scale_color() +
  geom_fruit(data = tree_qc_df, geom = geom_bar,
             mapping = aes(y = Bin, x = Contamination, fill = Contamination),
             orientation = "y", stat = "identity", position = "auto",
             grid.params = list(linetype = 1, size = 0.1), pwidth = 0.3, offset = 0.15,
             axis.params = list(axis = "x", text.size = 1.8, vjust = 1, nbreak = 3)) +
  scale_fill_gradient(name = "Contamination (%)", low = "#6a60a8", high = "#c4c2de",
                      guide = guide_legend(keyheight = 0.3, keywidth = 0.6, reverse = TRUE, order = 2)) +
  new_scale_fill() + new_scale_color() +
  geom_fruit(data = tree_qc_df, geom = geom_bar,
             mapping = aes(y = Bin, x = Strain_Het, fill = Strain_Het),
             orientation = "y", stat = "identity", position = "auto",
             grid.params = list(linetype = 1, size = 0.1), pwidth = 0.3, offset = 0.15,
             axis.params = list(axis = "x", text.size = 1.8, vjust = 1, nbreak = 3)) +
  scale_fill_gradient(name = "Strain Heterogeneity", low = "#c9543f", high = "#e9bab2",
                      guide = guide_legend(keyheight = 0.3, keywidth = 0.6, reverse = TRUE, order = 3)) +
  new_scale_fill() + new_scale_color() +
  geom_fruit(data = tree_qc_df, geom = geom_bar,
             mapping = aes(y = Bin, x = Pred_Genes, fill = Pred_Genes),
             orientation = "y", stat = "identity", position = "auto",
             grid.params = list(linetype = 1, size = 0.1), pwidth = 0.3, offset = 0.15,
             axis.params = list(axis = "x", text.size = 1.8, vjust = 1, nbreak = 2)) +
  scale_fill_gradient(name = "Predicted Genes", low = "#279854", high = "#a9dede",
                      guide = guide_legend(keyheight = 0.3, keywidth = 0.6, reverse = TRUE, order = 4)) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 9),
        legend.key.height = unit(4, units = "mm"),
        legend.key.width = unit(3, units = "mm"))
ggsave("article_profiling/1-taxonomy/plots/plot_biom_MAGs_tree_md.pdf", plot_tree_s15_md, height = 7, width = 9)
```

## Figure 1 & supplementary 1

```{r fig1}
# Merge plots using patchwork
fig_1a <- wrap_plots(plot_tree_s15_md, plot_mags_qc, plot_depth_sgbs_biom,
                     design = "\nAAAB\nAAAC", heights = c(1, 0.85))
#ggsave("article_profiling/fig_1a.pdf", fig_1a, height = 8, width = 15)
fig_1b <- wrap_plots(plot_depth_sgbs_both, plot_depth_spec_eco, plot_depth_kggbs_biom,
                     design = "\nAAABBCC\n")
#ggsave("article_profiling/fig_1b.pdf", fig_1b, height = 8, width = 15)

fig_1 <- wrap_plots(fig_1a, fig_1b, nrow = 2, heights = c(1, 0.85)) +
  plot_annotation(tag_levels = list(c("a", "b", "c", "d", "", "e", "f"))) &
  theme(plot.tag = element_text(face = "bold"))
ggsave("article_profiling/fig_1.pdf", fig_1, height = 16, width = 16)
#ggsave("article_profiling/fig_1.png", fig_1, height = 16, width = 16)
rm(fig_1a, fig_1b) #clean

#sup 1
fig_s1 <-  { plot_depth_top_phyla |
    (plot_depth_usgbs / plot_depth_spec + plot_layout(heights = c(0.85, 0.15))) } +
  plot_layout(ncol = 2, widths = c(0.25, 1)) +
  plot_annotation(tag_levels = list(c("a", "b", "c"))) &
  theme(plot.tag = element_text(size = 20, face = "bold"))
fig_s1[[2]][[1]] <- fig_s1[[2]][[1]] + theme(axis.title.y = element_text(vjust = 0))
fig_s1[[2]][[2]] <- fig_s1[[2]][[2]] + theme(axis.line.y.left = element_blank(),
                                             axis.ticks.y.left = element_blank())#,
                                             #axis.text.y = element_text(hjust = 2.2))
ggsave("article_profiling/fig_s1.pdf", fig_s1, height = 13, width = 22)
#ggsave("article_profiling/fig_s1.png", fig_s1, height = 13, width = 22)
```

## Section 2: Functional Annotation

### Pre-process matrix of KEGG KOs from KOfamScan

```{r processkos}
# Read output matrix from KOfamScan
kegg_annot <- read.csv("/mnt/beegfs/home/mrojas/DiGenomaLab/Systemix/analysis/kegg/merge_hits/kofamscan_hits.tsv", sep = "\t", check.names = FALSE)
colnames(kegg_annot)[2] <- "KO.definition"

# Remove uncharacterized and duplicated proteins
kegg_annot <- kegg_annot %>%
  dplyr::filter(!grepl("uncharacterized", KO.definition) == TRUE) %>%
  distinct(KO.definition, .keep_all = TRUE)

# Extract KO definitions and set KOs as row names
kegg_kos <- data.frame(KO = kegg_annot$KO, Definition = kegg_annot$KO.definition)
kegg_annot <- kegg_annot %>%
  dplyr::select(-c("KO.definition")) %>%
  data.frame(row.names = 1, check.names = FALSE)

# Set definitions for pathways of interest
kegg_paths <- kegg_kos %>%
  #sulf: to keep sulfur, sulfate, sulfane, sulfide, sulfo_
  #soxr: SoxR 2Fe-2S cluster
  #nitrogen: a negative marker (ubiquous; ex: MerR family)
  dplyr::filter(grepl("iron|sulf|soxr|copper|nitrogen",
                      Definition, ignore.case = TRUE) == TRUE) %>%
  dplyr::select(KO, Definition)

## FUNCTIONS ##
# Function that returns the pathway based on definitions
fx_setKOpathways <- function(def) {
  if (grepl("copper", def, ignore.case = TRUE) == TRUE)
    return("Copper")
  if (grepl("iron", def, ignore.case = TRUE) == TRUE)
    return("Iron")
  if (grepl("sulf|sox", def, ignore.case = TRUE) == TRUE)
    return("Sulfur")
  if (grepl("nitrogen", def, ignore.case = TRUE) == TRUE)
      return("Nitrogen")
  else
    return("Other")
}
# Function to define types of metabolism
fx_setKOmetabtypes <- function(def) {
  #oxidation, reduction
  #fixation, hydrolysis, production, utilization, transport
  #sulfane involved in oxidative processes, as well as dehydrogenases
  if (grepl("oxida|oxidi|hydrog|hydrox|sulfane", def, ignore.case = TRUE) == TRUE)
    return("Oxidation")
  if (grepl("reduc", def, ignore.case = TRUE) == TRUE)
    return("Reduction")
  if (grepl("utili", def, ignore.case = TRUE) == TRUE)
    return("Utilization")
  if (grepl("hydrolase", def, ignore.case = TRUE) == TRUE)
    return("Hydrolysis")
  if (grepl("transferase", def, ignore.case = TRUE) == TRUE)
    return("Transferase")
  if (grepl("transport|binding", def, ignore.case = TRUE) == TRUE)
    return("Transport")
  if (grepl("resistance", def, ignore.case = TRUE) == TRUE)
    return("Resistance")
  if (grepl("carrier|chaperone", def, ignore.case = TRUE) == TRUE)
    return("Carrier")
  if (grepl("regulat", def, ignore.case = TRUE) == TRUE)
    return("Regulator")
  if (grepl("lyase", def, ignore.case = TRUE) == TRUE)
    return("Lyase")
  else
    return("Other")
}
# Function that returns the ecosystem given a MAG
fx_getMAGecosystem <- function(mag) {
  if (grepl("spades", mag, ignore.case = TRUE) == TRUE) {
    return("Biomining")
  }
  else {
    mag_eco <- mags_metadata_bac %>%
      dplyr::mutate(bin = paste0(mag_name, ".", mag_id)) %>%
      dplyr::filter(bin == mag) %>%
      dplyr::select(ecosystem) %>%
      as.character()
    return(mag_eco)
  }
}
# Function that returns the genome size given a MAG
fx_getMAGlength <- function(mag) {
  mag_size <- mags_metadata_bac %>%
    dplyr::mutate(bin = paste0(mag_name, ".", mag_id)) %>%
    dplyr::filter(bin == mag) %>%
    dplyr::select(genome_size) %>%
    as.numeric()
  return(mag_size)
}
## END FUNCTIONS ##

# Populate both pathways and metabolism
kegg_ko_paths <- vector()
kegg_ko_metab <- vector()
for (i in 1:nrow(kegg_paths)) {
	kegg_ko_paths <- append(kegg_ko_paths, fx_setKOpathways(kegg_paths$Definition[i]))
	kegg_ko_metab <- append(kegg_ko_metab, fx_setKOmetabtypes(kegg_paths$Definition[i]))
}
kegg_paths["Pathway"] <- kegg_ko_paths
kegg_paths["Metabolism"] <- kegg_ko_metab
rm(kegg_ko_paths, kegg_ko_metab) #clean memory

# Keep only selected pathways from the annotation matrix
# (NOTE: this matrix will be used later for bubble plots)
kegg_filt <- kegg_annot[row.names(kegg_annot) %in% kegg_paths$KO, ]

# Annotate ecosystem and genome size for MAGs
kegg_mags <- colnames(kegg_filt) %>% as.data.frame()
colnames(kegg_mags) <- c("mag")

kegg_mags_eco <- vector()
kegg_mags_length <- vector()
for (i in 1:nrow(kegg_mags)) { #this loop takes a few secs to run!
	kegg_mags_eco <- append(kegg_mags_eco, fx_getMAGecosystem(kegg_mags$mag[i]))
	kegg_mags_length <- append(kegg_mags_length, fx_getMAGlength(kegg_mags$mag[i]))
}
kegg_mags["ecosystem"] <- kegg_mags_eco
kegg_mags["length"] <- kegg_mags_length
rm(kegg_mags_eco, kegg_mags_length) #clean memory

# Explore ecosystems..
table(kegg_mags$ecosystem)
kegg_mags %>% dplyr::filter(ecosystem == "Biomining") %>% dplyr::summarise(total = sum(length))
kegg_mags %>% dplyr::group_by(ecosystem) %>% dplyr::summarise(total = sum(length))
```

### Heatmap

```{r heatmap}
# Row-wise variance, then filter both matrix and pathways
kegg_hm_filt_var <- apply(kegg_filt, 1, var)
kegg_hm_data <- kegg_filt[kegg_hm_filt_var >= summary(kegg_hm_filt_var)[3], ] #median
kegg_hm_paths <- kegg_paths[kegg_paths$KO %in% row.names(kegg_hm_data), ]

# Normalize rows by z-score
kegg_hm_data <- apply(kegg_hm_data, 1, norm <- function(x){return (x - mean(x)) / sd(x)}) %>% t()

# Plot using ComplexHeatmap
kegg_hm_colors <- colorRamp2(c(-2, 0, 2), c("#5073AD", "white", "#D33D35"))

# Annotate rows: KO labels
#with high number of hits: row.names(kegg_hm_data[rowSums(kegg_hm_data) > 100, ])
# KOs with low variance (1/2 data %>% sort())
#kegg_hm_rows_lab <- row.names(kegg_hm_data[apply(kegg_filt_norm, 1, var) < 0.00015, ])
#kegg_hm_rows_at <- which(row.names(kegg_hm_data) %in% kegg_hm_rows_lab)
#kegg_hm_rows_kos <- HeatmapAnnotation(KOs = anno_mark(at = kegg_hm_rows_at,
#                                                      labels = kegg_hm_rows_lab,
#                                                      labels_gp = gpar(fontsize = 6)),
#                                      which = "row")

# Annotate rows: KO pathways
kegg_hm_rows_path <- c("Iron" = "#666666", "Sulfur" = "#DDAD3B", "Copper" = "#CA6627")
kegg_hm_rows_path <- list(Def = kegg_hm_paths$Pathway, Pathways = kegg_hm_rows_path)

kegg_hm_rows_ann_path <- HeatmapAnnotation(Pathways = kegg_hm_paths$Pathway,
                                           col = kegg_hm_rows_path, border = FALSE,
                                           show_annotation_name = FALSE,
                                           show_legend = TRUE, which = "row")

# Annotate rows: KO metabolism
kegg_hm_rows_metab <- setNames(brewer.pal(length(unique(kegg_hm_paths$Metabolism)), "Paired"), unique(kegg_hm_paths$Metabolism)) #Set3
kegg_hm_rows_metab <- sample(kegg_hm_rows_metab)
kegg_hm_rows_metab <- list(Met = kegg_hm_paths$Metabolism, Metabolism = kegg_hm_rows_metab)

kegg_hm_rows_ann_metab <- HeatmapAnnotation(Metabolism = kegg_hm_paths$Metabolism,
                                            col = kegg_hm_rows_metab, border = FALSE,
                                            show_annotation_name = FALSE,
                                            show_legend = TRUE, which = "row")

# Annotate cols: ecosystems
#match colors with mags_colors_eco*
kegg_hm_cols_mags <- setNames(brewer.pal(length(unique(kegg_mags$ecosystem)), "Paired"), unique(kegg_mags$ecosystem))
kegg_hm_cols_mags <- sample(kegg_hm_cols_mags)
kegg_hm_cols_mags <- list(Eco = kegg_mags$ecosystem, Ecosystem = kegg_hm_cols_mags)

kegg_hm_cols_ann_mags <- HeatmapAnnotation(Ecosystem = kegg_mags$ecosystem,
                                           col = kegg_hm_cols_mags, border = FALSE,
                                           show_annotation_name = FALSE,
                                           show_legend = TRUE, which = "column")

# Plot heatmap with annotations
pdf("plot_annot_heatmap.pdf", height = 9, width = 9)
kegg_hm_plot <- Heatmap(kegg_filt_norm, name = "Abundance",
                        row_title = "KEGG KOs", column_title = "Ecosystem MAGs",
                        show_row_names = FALSE, show_column_names = FALSE,
                        col = kegg_hm_colors, border = FALSE,
                        row_split = kegg_hm_paths$Pathway,
                        column_split = kegg_mags$ecosystem,
                        right_annotation = kegg_hm_rows_ann_metab,
                        left_annotation = kegg_hm_rows_ann_path,
                        top_annotation = kegg_hm_cols_ann_mags)
draw(kegg_hm_plot, merge_legend = TRUE)
while (!is.null(dev.list())) dev.off()
# Row names in the heatmap are not sorted!
#row.names(kegg_hm_data)[row_order(kegg_hm_plot)]
```

### PCA

```{r pca}
# NOTE: selected cycles (Fe, Cu, S) have low representation in row sums,
# so during filtration we need to consider them within the range of values.
# E.g., rows with low variances could be kept but they are also too many.
#kegg_pca_annot_row_var <- apply(kegg_annot, 1, var)
#kegg_pca_dat <- kegg_annot[kegg_pca_annot_row_var > summary(kegg_pca_annot_row_var)[3], ]
kegg_pca_dat <- kegg_hm_data

# Merge KO definitions
kegg_pca_kos <- data.frame(KO = row.names(kegg_pca_dat))
kegg_pca_kos <- left_join(kegg_pca_kos, kegg_kos, by = c("KO"))

# Pathways of interest to dataframe
kegg_pca_def <- vector()
for (i in 1:nrow(kegg_pca_kos)) {
	kegg_pca_def <- append(kegg_pca_def, fx_setKOpathways(kegg_pca_kos$Definition[i]))
}
kegg_pca_kos["Pathway"] <- kegg_pca_def

# Normalize data to z-score
kegg_pca_dat <- apply(kegg_pca_dat, 1, norm_pca <- function(x){return (x - mean(x)) / sd(x)}) %>% t()

# PCA based on SVD
set.seed(122)
kegg_pca <- prcomp(kegg_pca_dat, center = FALSE, scale. = FALSE)

# Print loadings (eigenvectors)
kegg_pca$rotation
kegg_pca$sdev

# Calculate variance and variance explained by each PC
kegg_pca_var <- kegg_pca$sdev ^ 2
kegg_pca_var_exp <- kegg_pca_var / sum(kegg_pca_var)

# Save as PCA object file
if (file.exists("kegg_pca.rds") == FALSE) {
  saveRDS(kegg_pca, "kegg_pca.rds")
}

# Screeplot
kegg_pca_scree_df <- data.frame(PC = 1:ncol(kegg_pca$rotation), Var = kegg_pca_var_exp)
kegg_pca_screeplot <- ggplot(kegg_pca_scree_df, aes(x = PC, y = Var)) +
  geom_bar(stat = "identity", fill = "#9cb8ca") +
  labs(x = "PC", y = "Variance Explained") +
  xlim(0, 50) + ylim(0, 0.06) +
  theme_minimal()
ggsave("plot_annot_pca_screeplot.pdf", kegg_pca_screeplot, height = 4, width = 5)

# Biplot
#kegg_pca_colors <- c("Copper" = "#CA6627", "Iron" = "#666666",
#                     "Sulfur" = "#DDAD3B", "Other" = "#e6f0f6")
kegg_pca_colors <- c("Copper" = "#CA6627", "Iron" = "#666666", "Sulfur" = "#DDAD3B")
kegg_pca_biplot <- ggbiplot(kegg_pca, choices = c(1, 2), groups = kegg_pca_kos$Pathway,
                            obs.scale = 1, var.scale = 1, var.axes = FALSE,
                            labels.size = 5, labels = "*", #alpha = 0,
                            ellipse = TRUE, circle = FALSE) +
  scale_color_manual(name = "Pathway", values = kegg_pca_colors,
                     guide = guide_legend(override.aes = aes(label = ""))) +
  geom_vline(xintercept = 0, linetype = 3) +
  geom_hline(yintercept = 0, linetype = 3) +
  xlim(-100, 140) +
  #geom_point(aes(colour = kegg_pca_kos$Pathway), size = 1.5) +
  theme(legend.direction = "horizontal", legend.position = "right") +
  theme_bw()
ggsave("plot_annot_pca_biplot.pdf", kegg_pca_biplot, height = 9, width = 9)


## PCAtools
# Try to separate biomining from other ecosystems using selected pathways
set.seed(221)
kegg_pca_obj <- pca(kegg_pca_dat, center = FALSE, scale = FALSE,
                    metadata = kegg_mags %>% data.frame(row.names = 1))

# Save PCA object
if (file.exists("kegg_pcatools.rds") == FALSE) {
  saveRDS(kegg_pca_obj, "kegg_pcatools.rds")
}

# Convert to a PCAtools object
# Same result, different y-scale!
#kegg_pca_obj <- list(rotated = data.frame(kegg_pca$rotation),
#                     loadings = data.frame(kegg_pca$x),
#                     variance = (kegg_pca_var_exp * 100) %>% set_names(colnames(kegg_pca$x)),
#                     sdev = kegg_pca$sdev,
#                     metadata = kegg_mags,
#                     xvars = kegg_pca_kos$KO,
#                     yvars = colnames(kegg_pca_mat),
#                     components = colnames(kegg_pca$x))
#row.names(kegg_pca_obj$rotated) <- kegg_pca_obj$yvars
#row.names(kegg_pca_obj$loadings) <- kegg_pca_obj$xvars
#names(kegg_pca_obj$variance) <- kegg_pca_obj$components
#class(kegg_pca_obj) <- "pca"

# Biplot
kegg_pca_obj_biplot <- biplot(kegg_pca_obj, showLoadings = FALSE, sizeLoadingsNames = FALSE,
                              drawConnectors = FALSE,  drawConnectorsLoadings = FALSE,
                              pointSize = 2, labSize = 1, legendPosition = "right",
                              colLegendTitle = "Ecosystem", colby = "ecosystem", 
                              colkey = mags_colors_eco, ellipse = TRUE,
                              xlim = c(-100, 70), ylim = c(-20, 10))
ggsave("plot_annot_pcatools_biplot.pdf", kegg_pca_obj_biplot, height = 9, width = 10)

# Pairsplot
kegg_pca_obj_pairs <- pairsplot(kegg_pca_obj, colby = "ecosystem",
                                colkey = mags_colors_eco, pointSize = 1,
                                components = getComponents(kegg_pca_obj, c(1:4)))
ggsave("plot_annot_pcatools_pairsplot.pdf", kegg_pca_obj_pairs, height = 7, width = 10)
```

### Bubble and boxplots

```{r bubble}
# Get unique ecosystems
kegg_bubble_eco <- unique(kegg_mags$ecosystem) %>% as.character()

# Take selected KOs and prepare dataset
kegg_bubble <- row.names(kegg_filt) %>%
  as.data.frame() %>%
  slice(rep(1:n(), each = length(kegg_bubble_eco)))
colnames(kegg_bubble) <- c("KO")
kegg_bubble["ecosystem"] <- c("")
kegg_bubble$ecosystem <- rep(kegg_bubble_eco, nrow(kegg_filt))

# Plot design/overview:
# -split KOs per Cu, Fe and S (N is control)
# -split x-axis per ecosystems, KO definitions in y-axis
# -sum KO counts across ecosystems
# -normalize them taking the genome size
# -use statistical analysis (anova, groups)

# Function that counts the matching KOs in each ecosystem
# Also, retrieves the bins size in order to normalize counts later
fx_countKOsByEco <- function(ko) {
  #retrieve all bins in the filtered matrix for the given KO
  #this row contains the counts across all ecosystems
  row <- kegg_filt[row.names(kegg_filt) == ko, ]
  #vectors to populate
  eco <- vector()
  len <- vector()
  cou <- vector()
  #get the ecosystem, length and counts for each bin
  for (j in 1:ncol(row)) {
    e <- kegg_mags[kegg_mags$mag == colnames(kegg_filt)[j], ] %>%
      dplyr::select(ecosystem) %>% as.character()
    eco <- append(eco, e)
    l <- kegg_mags[kegg_mags$mag == colnames(kegg_filt)[j], ] %>%
      dplyr::select(length) %>% as.numeric()
    len <- append(len, l) #length
    cou <- append(cou, row[1, j]) #counts
  }
  #vectors to dataframe
  df <- data.frame(ecosystem = eco, count = cou, length = len)
  #sum counts and length by ecosystem
  df <- df %>% dplyr::group_by(ecosystem) %>%
    dplyr::summarise(across(everything(), sum))
  return(df)
}

# Count total KOs and length in all across ecosystems for selected pathways
# (time-consuming loop: runtime ~12 mins for rows in kegg_paths*)
kegg_bubble_counts <- tibble(ecosystem = character(), count = numeric(), length = numeric())
kos_df <- vector()
eco_df <- vector()
cou_df <- vector()
for (i in seq(1, nrow(kegg_bubble), length(kegg_bubble_eco))) {
  #current KO -> loop step of 10 KOs
  #print(sprintf("%i -> KO: %s", i, kegg_bubble[i, 1]))
  kos_df <- fx_countKOsByEco(kegg_bubble[i, 1])
  #merge ecosystem counts sorted from our list
  eco_df <- kegg_bubble_eco %>% as.data.frame()
  colnames(eco_df) <- c("ecosystem")
  #save counts
  cou_df <- left_join(eco_df, kos_df, by = c("ecosystem"))
  kegg_bubble_counts <- rbind(kegg_bubble_counts, cou_df)
}
# Add column for counts
kegg_bubble <- cbind(kegg_bubble, counts = kegg_bubble_counts[, 2])
kegg_bubble <- cbind(kegg_bubble, length = kegg_bubble_counts[, 3])
rm(kos_df, eco_df, cou_df) #clean memory

# Normalize counts: length to mb and get counts per mb
kegg_bubble <- kegg_bubble %>%
  dplyr::mutate(length_mb = round(length / 1000000, 3)) %>%
  dplyr::mutate(count_norm = round(counts / length_mb, 3))
#explore
kegg_bubble %>% dplyr::group_by(ecosystem) %>% dplyr::summarise(n = sum(count_norm))

# Combine counts and pathways datasets:
# -get definitions and metabolism from kegg_paths
# -merge df with counts per ecosystem
# -remove KOs from biomining with low counts

# Copper pathway
kegg_bubble_cu <- kegg_paths %>%
  dplyr::filter(Pathway == "Copper") %>%
  dplyr::select(KO, Definition, Metabolism)
kegg_bubble_cu <- left_join(kegg_bubble_cu, kegg_bubble, by = c("KO"))
#filter
kegg_bubble_keep <- kegg_bubble_cu %>%
  dplyr::filter(ecosystem == "Biomining" & count_norm > 0) %>%
  dplyr::select(KO)
kegg_bubble_cu <- kegg_bubble_cu[kegg_bubble_cu$KO %in% kegg_bubble_keep$KO, ]

# Iron pathway
kegg_bubble_fe <- kegg_paths %>%
  dplyr::filter(Pathway == "Iron") %>%
  dplyr::select(KO, Definition, Metabolism)
kegg_bubble_fe <- left_join(kegg_bubble_fe, kegg_bubble, by = c("KO"))
#filter
kegg_bubble_keep <- kegg_bubble_fe %>%
  dplyr::filter(ecosystem == "Biomining" & count_norm > 0.01) %>%
  dplyr::select(KO)
kegg_bubble_fe <- kegg_bubble_fe[kegg_bubble_fe$KO %in% kegg_bubble_keep$KO, ] %>%
  dplyr::filter(!KO == "K02013" & !KO == "K02014") #outliers

# Sulfur pathway
kegg_bubble_s <- kegg_paths %>%
  dplyr::filter(Pathway == "Sulfur") %>%
  dplyr::select(KO, Definition, Metabolism)
kegg_bubble_s <- left_join(kegg_bubble_s, kegg_bubble, by = c("KO"))
#filter
kegg_bubble_keep <- kegg_bubble_s %>%
  dplyr::filter(ecosystem == "Biomining" & count_norm > 0.05) %>%
  dplyr::select(KO)
kegg_bubble_s <- kegg_bubble_s[kegg_bubble_s$KO %in% kegg_bubble_keep$KO, ]

# Sort ecosystems by median
anova_sort_cu <- kegg_bubble_cu %>%
  dplyr::group_by(ecosystem) %>%
  dplyr::summarise(n = median(count_norm)) %>%
  dplyr::arrange(desc(n)) %>%
  as.data.frame()
anova_sort_fe <- kegg_bubble_fe %>%
  dplyr::group_by(ecosystem) %>%
  dplyr::summarise(n = median(count_norm)) %>%
  dplyr::arrange(desc(n)) %>%
  as.data.frame()
anova_sort_s <- kegg_bubble_s %>%
  dplyr::group_by(ecosystem) %>%
  dplyr::summarise(n = median(count_norm)) %>%
  dplyr::arrange(desc(n)) %>%
  as.data.frame()

# We need to re-sort the default colors..
mags_ecosystem <- kegg_bubble_eco %>% sort()
mags_colors <- tibble(ecosystem = mags_ecosystem, colors = mags_colors_eco)
anova_sort_cu <- left_join(anova_sort_cu, mags_colors, by = c("ecosystem"))
anova_sort_fe <- left_join(anova_sort_fe, mags_colors, by = c("ecosystem"))
anova_sort_s <- left_join(anova_sort_s, mags_colors, by = c("ecosystem"))

# Boxplots
anova_boxplot_cu <- ggboxplot(kegg_bubble_cu, x = "ecosystem", y = "count_norm",
                           color = "ecosystem", palette = anova_sort_cu$colors,
                           order = anova_sort_cu$ecosystem, add = "jitter", legend = "none") +
  geom_hline(yintercept = mean(kegg_bubble_cu$count_norm), linetype = 2, linewidth = 0.5) +
  labs(x = "Ecosystem", y = "Copper pathway (KOs/Mb)") +
  rotate_x_text(angle = 45) + ylim(c(0, 0.3)) +
  #stat_summary(fun = mean, geom = "point", shape = 18, size = 2.5, color = "#FC4E07") +
  stat_compare_means(method = "anova", label.y = 0.3, size = 5) + #add global annova p-value
  stat_compare_means(label = "p.signif", method = "t.test", ref.group = ".all.", hide.ns = TRUE) +
  theme(axis.text = element_text(size = 14), axis.title = element_text(size = 16))
ggsave("article_profiling/4-mags-smag/2-plots-annotation/bubble/plot_annot_boxplot_copper.pdf", anova_boxplot_cu, height = 7, width = 7)

anova_boxplot_fe <- ggboxplot(kegg_bubble_fe, x = "ecosystem", y = "count_norm",
                           color = "ecosystem", palette = anova_sort_fe$colors,
                           order = anova_sort_fe$ecosystem, add = "jitter", legend = "none") +
  geom_hline(yintercept = mean(kegg_bubble_fe$count_norm), linetype = 2, linewidth = 0.5) +
  labs(x = "Ecosystem", y = "Iron pathway (KOs/Mb)") +
  rotate_x_text(angle = 45) + ylim(c(0, 0.4)) +
  stat_compare_means(method = "anova", label.y = 0.4, size = 5) + #add global annova p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.", hide.ns = TRUE, label.y = 0.35) +
  theme(axis.text = element_text(size = 14), axis.title = element_text(size = 16))
ggsave("article_profiling/4-mags-smag/2-plots-annotation/bubble/plot_annot_boxplot_iron.pdf", anova_boxplot_fe, height = 7, width = 7)

anova_boxplot_s <- ggboxplot(kegg_bubble_s, x = "ecosystem", y = "count_norm",
                             color = "ecosystem", palette = anova_sort_s$colors,
                             order = anova_sort_s$ecosystem, add = "jitter", legend = "none") +
  geom_hline(yintercept = mean(kegg_bubble_s$count_norm), linetype = 2, linewidth = 0.5) +
  labs(x = "Ecosystem", y = "Sulfur pathway (KOs/Mb)") +
  rotate_x_text(angle = 45) + ylim(c(0, 0.8)) +
  stat_compare_means(method = "anova", label.y = 0.8, size = 5) + #add global annova p-value
  stat_compare_means(label = "p.signif", method = "t.test", ref.group = ".all.",
                     hide.ns = TRUE, label.y = 0.75, size = 6) +
  theme(axis.text = element_text(size = 14), axis.title = element_text(size = 16))
ggsave("article_profiling/4-mags-smag/2-plots-annotation/bubble/plot_annot_boxplot_sulfur.pdf", anova_boxplot_s, height = 7, width = 7)

# Bubble plot with key KOs from biomining
#multicopper, copper resistance (B,C,D), menaquinol
#UDP−sulfoquinovose, high−affinity iron transporter, thiosulfate dehydrogenase
#sulfide:quinone oxidoreductase, sulfate adenylyltransferase, mononucleotide sulfurtransferase, 
#heterodisulfide C2, sulfolactate sulfo-lyase subunit, CysO sulfur-carrier
kegg_biom_key <- c("K22552", "K07233", "K07156", "K07245", "K03886",
                   "K06118", "K07243", "K19713", "K16937",
                   "K17218", "K00958", "K06864", 
                   "K03390", "K01023", "K16845", "K16846", "K21140", "K21148")

kegg_bubbles_plot <- kegg_bubble %>%
  filter(KO %in% kegg_biom_key) %>%
  left_join(kegg_paths, by = c("KO")) %>% #add definitions, then plot
  ggplot(aes(x = reorder(ecosystem, -count_norm),
             y = Definition, size = count_norm, color = ecosystem)) +
  geom_point(alpha = 0.5) +
  scale_size(range = c(0, 10)) +
  scale_y_discrete(position = "right") +
  scale_color_manual(values = mags_colors_eco, guide = "none") +
  labs(title = "", x = "Ecosystem", y = "") +
  guides(size = guide_legend(title = "Norm. counts")) +
  theme_classic() +
  theme(legend.title = element_text(size = 12), legend.position = "top",
        axis.text.x = element_text(size = 14, angle = 90, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 14), axis.title.x = element_text(size = 16),
        axis.line.y.right = element_blank(), axis.ticks.y.right = element_blank())
ggsave("article_profiling/4-mags-smag/2-plots-annotation/bubble/plot_annot_bubbles.pdf", kegg_bubbles_plot, height = 7, width = 12)

# ANOVA test:
# H0 the means are equal for ecosystems (if p < 0.05, we take Ha)
#anova_cu <- aov(count_norm ~ ecosystem, data = kegg_bubble_cu)
#summary(anova_cu) #p = 0.0866
#anova_iron <- aov(counts ~ ecosystem, data = kegg_bubble_iron)
#summary(anova_iron) #p = 0.00912 ** (signif. 0.01)
#anova_sulfur <- aov(counts ~ ecosystem, data = kegg_bubble_sulfur)
#summary(anova_sulfur) #p = 6.3e-10 *** (signif. 0.001)

## Calculate mean and sd between ecosystems
#anova_copper_stats <- kegg_bubble_copper %>%
#  group_by(ecosystem) %>%
#  dplyr::summarise(mean = mean(counts), sd = sd(counts))
#anova_iron_stats <- kegg_bubble_iron %>%
#  group_by(ecosystem) %>%
#  dplyr::summarise(mean = mean(counts), sd = sd(counts))
#anova_sulfur_stats <- kegg_bubble_sulfur %>%
#  group_by(ecosystem) %>%
#  dplyr::summarise(mean = mean(counts), sd = sd(counts))
#
## FISHER LSD test: inspect which group means are different (ie. tukey)
## H0: ecosystems are independent
#anova_copper_test <- LSD.test(anova_copper, "ecosystem")
##print(anova_copper_test)
#anova_copper_test <- anova_copper_test$groups %>%
#  as.data.frame() %>%
#  rownames_to_column("ecosystem")
#anova_iron_test <- LSD.test(anova_iron, "ecosystem")
#anova_iron_test <- anova_iron_test$groups %>%
#  as.data.frame() %>%
#  rownames_to_column("ecosystem")
#anova_sulfur_test <- LSD.test(anova_sulfur, "ecosystem")
#anova_sulfur_test <- anova_sulfur_test$groups %>%
#  as.data.frame() %>%
#  rownames_to_column("ecosystem")

# Box plots -> from checking use reorder()* instead fct_reorder()*!
#anova_copper_plot <- ggplot(kegg_bubble_copper, aes(x = reorder(ecosystem, -counts),
#                                                    y = counts, color = ecosystem)) +
#  geom_boxplot() +
#  scale_color_manual(values = mags_colors_eco, guide = "none") +
#  geom_text(data = anova_copper_test, aes(label = groups),
#            y = -1.5, vjust = -0.5, size = 3, color = "black") +
#  labs(x = "Ecosystem", y = "KO counts for copper") +
#  theme_bw() +
#  theme(panel.grid.major = element_blank(),
#        panel.grid.minor = element_blank(),
#        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
#ggsave("plot_annot_bubble_anova_copper.pdf", anova_copper_plot, height = 7, width = 7)
#
#anova_iron_plot <- ggplot(kegg_bubble_iron, aes(x = reorder(ecosystem, -counts),
#                                                y = counts, color = ecosystem)) +
#  geom_boxplot() +
#  scale_color_manual(values = mags_colors_eco, guide = "none") +
#  geom_text(data = anova_iron_test, aes(label = groups),
#            y = -5, vjust = -0.5, size = 3, color = "black") +
#  labs(x = "Ecosystem", y = "KO counts for iron") +
#  theme_bw() +
#  theme(panel.grid.major = element_blank(),
#        panel.grid.minor = element_blank(),
#        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
#ggsave("plot_annot_bubble_anova_iron.pdf", anova_iron_plot, height = 7, width = 7)
#
#anova_sulfur_plot <- ggplot(kegg_bubble_sulfur, aes(x = reorder(ecosystem, -counts),
#                                                    y = counts, color = ecosystem)) +
#  geom_boxplot() +
#  scale_color_manual(values = mags_colors_eco, guide = "none") +
#  geom_text(data = anova_sulfur_test, aes(label = groups),
#            y = -4, vjust = -0.5, size = 3, color = "black") +
#  labs(x = "Ecosystem", y = "KO counts for sulfur") +
#  theme_bw() +
#  theme(panel.grid.major = element_blank(),
#        panel.grid.minor = element_blank(),
#        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
#ggsave("plot_annot_bubble_anova_sulfur.pdf", anova_sulfur_plot, height = 7, width = 7)

# Bubble plots for each pathway
kegg_bubble_plot_cu <- ggplot(kegg_bubble_cu,
                              aes(x = reorder(ecosystem, -count_norm),
                                  y = Definition, size = count_norm, color = ecosystem)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = mags_colors_eco, guide = "none") +
  #scale_y_discrete(expand = c(0.06, 0)) + #expand y-axis 6% in both sides
  #geom_text(data = anova_copper_test, aes(x = ecosystem, y = 15.4, label = groups),
  #          vjust = -0.5, size = 2.5, inherit.aes = FALSE) +
  #geom_hline(yintercept = 15.4, linetype = 1, linewidth = 0.1) +
  #scale_x_discrete(expand = expansion(add = 0.85)) + #expand x-axis in units at both sides
  labs(title = "", subtitle = "p = 0.087",
       x = "Ecosystem", y = "KO definitions for copper") +
  guides(size = guide_legend(title = "Norm. counts")) +
  theme_classic() +
  theme(legend.title = element_text(size = 9), legend.position = "right",
        plot.subtitle = element_text(size = 8, hjust = 0.5, face = "italic"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
ggsave("plot_annot_bubble_copper.pdf", kegg_bubble_plot_cu, height = 7, width = 9)

kegg_bubble_plot_fe <- ggplot(kegg_bubble_fe,
                              aes(x = reorder(ecosystem, -count_norm),
                                  y = Definition, size = count_norm, color = ecosystem)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = mags_colors_eco, guide = "none") +
  #scale_y_discrete(expand = c(0.045, 0)) + #expand y-axis 4.5% in both sides
  #geom_text(data = anova_iron_test,
  #          aes(x = ecosystem, y = 20.4, label = groups),
  #          vjust = -0.5, size = 2.5, inherit.aes = FALSE) +
  #geom_hline(yintercept = 20.4, linetype = 1, linewidth = 0.1) +
  labs(title = "", subtitle = "p = 0.99",
       x = "Ecosystem", y = "KO definitions for iron") +
  guides(size = guide_legend(title = "Norm. counts")) +
  theme_classic() +
  theme(legend.title = element_text(size = 9), legend.position = "right",
        plot.subtitle = element_text(size = 8, hjust = 0.5, face = "italic"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
ggsave("plot_annot_bubble_iron.pdf", kegg_bubble_plot_fe, height = 7, width = 8)

kegg_bubble_plot_s <- ggplot(kegg_bubble_s,
                             aes(x = reorder(ecosystem, -count_norm),
                                 y = Definition, size = count_norm, color = ecosystem)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = mags_colors_eco, guide = "none") +
  #scale_y_discrete(expand = c(0.03, 0)) + #expand y-axis in both sides 3%
  #geom_text(data = anova_sulfur_test,
  #          aes(x = ecosystem, y = 41.6, label = groups),
  #          vjust = -0.5, size = 2.5, inherit.aes = FALSE) +
  #geom_hline(yintercept = 41.6, linetype = 1, linewidth = 0.1) +
  labs(title = "", subtitle = "p = 0.062",
       x = "Ecosystem", y = "KO definitions for sulfur") +
  guides(size = guide_legend(title = "Norm. counts")) +
  theme_classic() +
  theme(legend.title = element_text(size = 9), legend.position = "right",
        plot.subtitle = element_text(size = 8, hjust = 0.5, face = "italic"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
ggsave("plot_annot_bubble_sulfur.pdf", kegg_bubble_plot_s, height = 8, width = 9)
```

## Figure 2 & supplementary 2

```{r fig2}
# Merge plots using patchwork
fig_2 <- wrap_plots(anova_boxplot_cu, anova_boxplot_fe, anova_boxplot_s, design = "AABBCC") +
  plot_annotation(tag_levels = list(c("a", "b", "c"))) &
  theme(plot.tag = element_text(size = 22, face = "bold"))
ggsave("article_profiling/fig_2.pdf", fig_2, height = 7, width = 19)
ggsave("article_profiling/fig_2.png", fig_2, height = 7, width = 19)

#sup 2
fig_s2 <- wrap_plots(kegg_bubbles_plot, design = "AAA") +
  plot_annotation(tag_levels = list(c("a", "b", "c"))) &
  theme(plot.tag = element_text(size = 20, face = "bold"))
ggsave("article_profiling/fig_s2a.pdf", fig_s2, height = 7, width = 12)
#ggsave("article_profiling/fig_s2a.png", fig_s2, height = 7, width = 12)
```

## Section 3: Evolutionary analysis

### dN/dS

```{r snps}
# Buildref for 1 bin (testing for bacteria species..)
#https://github.com/im3sanger/dndscv/issues/62
#buildref(cdsfile = "article_profiling/5-SNPs/dndscv/buildref/cds_node_25854.txt",
#         genomefile = "article_profiling/5-SNPs/dndscv/buildref/node_25854.fa",
#         outfile = "article_profiling/5-SNPs/dndscv/buildref/testref.rda")
#load("article_profiling/5-SNPs/dndscv/buildref/testref.rda")

# Buildref for all bins (very slow! ~2 hrs in building object)
#buildref(cdsfile = "article_profiling/5-SNPs/dndscv/buildref/cds_table.txt",
#         genomefile = "article_profiling/5-SNPs/dndscv/buildref/S15_bins.fa",
#         outfile = "article_profiling/5-SNPs/dndscv/buildref/S15_refcds.rda")

# Inspect refdb
#load("article_profiling/5-SNPs/dndscv/buildref/S15_refcds.rda")
#print(RefCDS[[1]])
#rm(gr_genes, RefCDS) #clean memory

# Get annotations (parsed from the VCF generated with SnpEff)
#sampleID chr pos ref mut
snp_mutations <- read.csv("article_profiling/5-SNPs/dndscv/run/S15_mutations.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# Run dndscv (runtime ~8 min, high memory usage) and save object
#snp_dndsout <- dndscv(snp_mutations, refdb = "article_profiling/5-SNPs/dndscv/buildref/S15_refcds.rda",
#                      max_muts_per_gene_per_sample = Inf, max_coding_muts_per_sample = Inf,
#                      cv = NULL, outmats = TRUE)
#save(snp_dndsout, file = "article_profiling/5-SNPs/dndscv/run/S15_dndsout.rda")
#save(substmodel, file = "article_profiling/5-SNPs/dndscv/run/S15_substmodel.rda")
#load("article_profiling/5-SNPs/dndscv/run/S15_dndsout.rda")
#load("article_profiling/5-SNPs/dndscv/run/S15_substmodel.rda")
rm(snp_mutations) #clean memory

# Explore output object..
#https://github.com/im3sanger/dndscv/issues/6
#https://github.com/im3sanger/dndscv/issues/64
head(snp_dndsout$sel_cv) #dN/dS ratios at gene-level
head(snp_dndsout$annotmuts) #list of annotated mutations
print(snp_dndsout$globaldnds) #driver globals

# Save matrix with dN/dS: driver discovery or genes under positive selection (PS)
snp_selcv <- snp_dndsout$sel_cv
rm(snp_dndsout) #clean memory

# Group global dN/dS ratios:
#https://github.com/im3sanger/dndscv/issues/63
snp_global_ps <- snp_selcv %>%
  dplyr::filter(wmis_cv > 1) %>%
  dplyr::summarise(genes = n())
snp_global_ps$label = "dN/dS > 1"
snp_global_ns <- snp_selcv %>%
  dplyr::filter(wmis_cv < 1 & wmis_cv != 0) %>%
  dplyr::summarise(genes = n())
snp_global_ns$label = "dN/dS < 1"
snp_global_dnds <- rbind(snp_global_ps, snp_global_ns)
snp_global_dnds <- snp_global_dnds %>%
  dplyr::mutate(perc = round(genes * 100 / nrow(snp_selcv), 2)) %>% #add %
  dplyr::mutate(perc2 = paste0(genes, " (", perc, "%)")) #label to plot
rm(snp_global_ps, snp_global_ns) #clean memory

# Barplot with global dN/dS ratios (not used- for reference only*)
snp_plot_dndsglob <- ggplot(snp_global_dnds, aes(x = genes, y = reorder(label, genes), fill = label)) +
  geom_col(position = "dodge2", fill = "#4c6194") +
  geom_text(aes(label = perc2), vjust = 0.5, hjust = -0.1, size = 2) +
  scale_x_continuous(limits = c(0, 16000)) +
  labs(x = "Number of genes", y = "") +
  theme_classic() +
  theme(legend.position = "none")
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_global.pdf", snp_plot_dndsglob, height = 2, width = 5)

# Filter genes under PS:
#https://github.com/im3sanger/dndscv/issues/56
#wmis_cv: dN/dS for missense mutations
#pmis_cv: p-value (signif. < 0.05) is expected to occur by chance under neutrality (null hypothesis)
snp_psgenes <- snp_selcv %>%
  dplyr::filter(wmis_cv > 1) %>%
  dplyr::mutate(dndsr = round((n_mis / n_syn), 1))

# Get CDS annotations (pre-processed from Bakta results)
snp_refcds <- read.csv("article_profiling/5-SNPs/dndscv/buildref/cds_table.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# Load contig to bin matrix
snp_bins <- read.csv("article_profiling/5-SNPs/dndscv/results/S15_DASTool_contig2bin.tsv", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# Load gene names (parsed from genes.gff file)
snp_genids <- read.csv("article_profiling/5-SNPs/dndscv/results/gene_id_name.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

## FUNCTIONS ##
# Function that returns the contig where a gene was found
fx_dndsGene2Contig <- function(g) {
  c <- snp_refcds %>%
    dplyr::filter(gene.name == g) %>%
    dplyr::select(chr) %>%
    as.character()
  return(c)
}
# Function that returns the bin given a contig
fx_dndsContig2Bin <- function(c) {
  b <- snp_bins %>%
    dplyr::filter(contig == c) %>%
    dplyr::select(bin) %>%
    as.character()
  return(b)
}
# Function that gets gene names given an id
fx_dndsGeneName <- function(g) {
  g <- snp_genids %>%
    dplyr::filter(gene.id == g) %>%
    dplyr::select(gene.name) %>%
    as.character()
  return(g)
}
# Function that returns a row with the taxonomic classification of a bin
fx_dndsTaxaBin <- function(b) {
  t <- mag_samples %>%
    dplyr::mutate(bin = paste0(mag_name, ".", mag_id)) %>%
    dplyr::filter(bin == b) %>%
    dplyr::select(phylum, genus, species, completeness, contamination, predicted_genes)
  return(t)
}
# Function that process the product tag of NS genes
fx_dndsNsPathway <- function(p) {
  if (grepl("copper", p, ignore.case = TRUE) == TRUE)
    return("Copper")
  if (grepl("iron", p, ignore.case = TRUE) == TRUE)
    return("Iron")
  if (grepl("nickel", p, ignore.case = TRUE) == TRUE)
    return("Nickel")
  if (grepl("sulfur", p, ignore.case = TRUE) == TRUE)
    return("Sulfur")
  if (grepl("nitrogen", p, ignore.case = TRUE) == TRUE)
    return("Nitrogen")
}
# Function that gets dN/dS ratio and p-value for NS genes
fx_dndsNsStats <- function(g) {
  s <- snp_selcv %>%
    dplyr::filter(gene_name == g) %>%
    dplyr::select(gene_name, wmis_cv, pmis_cv)
  return(s)
}
## END FUNCTIONS ##

# Populate dataset with additional attributes per gene
snp_gene2contig <- vector()
snp_contig2bin <- vector()
snp_genenames <- vector()
for (i in 1:nrow(snp_psgenes)) {
	snp_gene2contig <- append(snp_gene2contig, fx_dndsGene2Contig(snp_psgenes$gene_name[i]))
	snp_contig2bin <- append(snp_contig2bin, fx_dndsContig2Bin(snp_gene2contig[i]))
	snp_genenames <- append(snp_genenames, fx_dndsGeneName(snp_psgenes$gene_name[i]))
}
snp_psgenes["contig"] <- snp_gene2contig
snp_psgenes["bin"] <- snp_contig2bin
snp_psgenes["name"] <- snp_genenames
rm(snp_refcds, snp_bins, snp_genids, snp_gene2contig, snp_contig2bin, snp_genenames) #clean memory

# Group PS genes to add taxa & QC
snp_dnds <- snp_psgenes %>%
  dplyr::group_by(bin) %>%
  dplyr::summarise(num = n()) %>%
  dplyr::arrange(desc(num))

snp_taxabins <- tibble(phylum = character(), genus = character(), species = character(),
                       completeness = numeric(), contamination = numeric(), predicted_genes = numeric())
for (i in 1:nrow(snp_dnds)) {
  #print(sprintf("%i -> bin: %s", i, fx_dndsTaxaBin(snp_dnds$bin[i])))
  snp_taxabins <- rbind(snp_taxabins, fx_dndsTaxaBin(snp_dnds$bin[i]))
}
snp_dnds["phylum"] <- snp_taxabins$phylum
snp_dnds["genus"] <- snp_taxabins$genus
snp_dnds["species"] <- snp_taxabins$species
snp_dnds["completeness"] <- snp_taxabins$completeness
snp_dnds["contamination"] <- snp_taxabins$contamination
snp_dnds["predicted_genes"] <- snp_taxabins$predicted_genes
rm(snp_taxabins) #clean memory

# Set unknown genus/species (uGGBs and uSGBs)
snp_dnds <- snp_dnds %>%
  dplyr::mutate(genus = ifelse(genus == "", "uGGB", genus)) %>%
  dplyr::mutate(species = ifelse(species == "", "uSGB", species))

# Subset PS genes to plot mutations -> synonymous/missense (non-syn.)/nonsense
#discard hypothetical proteins (HPs: according to name)
#genes with at least 4 SNPs
#mark genes reaching significance (q < 0.05)
#mark genes: KJCAEPPB_04437 -> nickel pathway; NEBJFEFK_01910 -> sulfur pathway
snp_psgenes_sub <- snp_psgenes %>%
  dplyr::filter(nchar(name) < 14) %>%
  dplyr::mutate(snps = n_syn + n_mis + n_non) %>%
  dplyr::filter(snps >= 4) %>%
  dplyr::mutate(signif = ifelse(pmis_cv < 0.05, "*", "")) %>%
  dplyr::mutate(path = case_when(gene_name == "KJCAEPPB_04437" ~ "Ni",
                                 gene_name == "NEBJFEFK_01910" ~ "S", TRUE ~ "")) %>%
  distinct(name, .keep_all = TRUE) %>% #remove repeated names
  pivot_longer(cols = c(n_syn, n_mis, n_non), names_to = "mut", values_to = "n_mut")

# Save PS gene names and ids to file
if (file.exists("article_profiling/5-SNPs/dndscv/results/ps_genes.txt") == FALSE) {
  write.table(snp_psgenes_sub %>% dplyr::select(gene_name) %>% distinct(gene_name), "article_profiling/5-SNPs/dndscv/results/ps_genes.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}
if (file.exists("article_profiling/5-SNPs/dndscv/results/kofam/ps_genes_ids.txt") == FALSE) {
  write.table(unique(snp_psgenes_sub$name), "article_profiling/5-SNPs/dndscv/results/kofam/ps_genes_ids.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}

# Plot global driver genes by MAG/species
snp_plot_psbins <- snp_dnds %>%
  dplyr::mutate(bin = str_replace_all(bin, "SPAdes-MaxBin2Refined-", "")) %>% #trim bin names
  dplyr::mutate(bin = str_replace_all(bin, "SPAdes-MetaBAT2Refined-", "")) %>%
  ggplot(aes(x = num, y = reorder(bin, num), fill = bin)) +
  geom_bar(stat = "identity", fill = "#e31a1c") +
  geom_text(aes(label = num), vjust = 0.5, hjust = -0.25, size = 4) +
  labs(x = "Number of genes under PS", y = "MAG Bin") +
  scale_x_continuous(limits = c(0, 255)) +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text = element_text(size = 16),
        axis.title = element_text(size = 18), strip.text.x = element_text(size = 14)) +
  facet_grid(.~species)
#ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_psbins.pdf", snp_plot_psbins, height = 11, width = 12)
#add completeness & contamination barplots
snp_plot_psbins2 <- snp_dnds %>%
  ggplot(aes(x = completeness, y = reorder(bin, num), fill = bin)) +
  geom_col(fill = "#3c4d76") +
  geom_vline(xintercept = 100, linetype = 1, linewidth = 0.4) +
  labs(x = "Completeness", y = "") +
  theme_classic() +
  theme(axis.text.y = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
snp_plot_psbins3 <- snp_dnds %>%
  ggplot(aes(x = contamination, y = reorder(bin, num), fill = bin)) +
  geom_col(fill = "#763c4d") +
  geom_vline(xintercept = max(snp_dnds$contamination), linetype = 1, linewidth = 0.4) +
  labs(x = "Contamination", y = "") +
  theme_classic() +
  theme(axis.text.y = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
#merge
snp_plot_psbins <- wrap_plots(snp_plot_psbins, snp_plot_psbins2, snp_plot_psbins3,
                              nrow = 1, ncol = 3, widths = c(1, 0.15, 0.15))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_psbins.pdf", snp_plot_psbins, height = 11, width = 17)
rm(snp_plot_psbins2, snp_plot_psbins3) #clean memory

# Lollipop plot of PS genes
snp_plot_psgenes <- snp_psgenes_sub %>%
  dplyr::filter(wmis_cv > 1.4) %>% #cutoff
  ggplot(aes(x = wmis_cv, y = reorder(name, wmis_cv)), size = wmis_cv) +
  geom_segment(aes(x = 0, xend = wmis_cv, y = name), color = "gray") +
  geom_point(alpha = 0.7, size = 2, color = "#e31a1c") +
  geom_text(aes(label = signif, x = -0.15), size = 6.5, color = "#3c4d76") +
  geom_text(aes(label = path, x = -0.15), size = 5.5, color = "#3c4d76") +
  geom_vline(xintercept = 1, linetype = 2, linewidth = 0.4) +
  geom_vline(xintercept = max(snp_psgenes_sub$wmis_cv), linetype = 1, linewidth = 0.4) +
  labs(x = "PS strength", y = "Gene name") +
  theme_classic() +
  theme(axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
#ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_psgenes.pdf", snp_plot_psgenes, height = 12, width = 7)

# Plot PS genes mutations per type, then merge
snp_plot_muts <- snp_psgenes_sub %>%
  dplyr::filter(wmis_cv > 1.4) %>% #cutoff
  ggplot(aes(x = n_mut, y = reorder(name, wmis_cv), fill = mut)) +
  geom_col(position = position_stack(reverse = TRUE)) +
  scale_fill_manual(name = "", labels = c("mis", "non", "syn"),
                    values = c("#5db4b2", "#b687b6", "#2374a4")) +
  labs(x = "Number of mutations", y = "") +
  theme_classic() +
  theme(axis.text.y = element_blank(), axis.line.y = element_blank(),
        legend.position = c(0.88, 0.15), strip.placement = "outside",
        axis.ticks.y = element_blank(), legend.text = element_text(size = 16),
        axis.text = element_text(size = 16), axis.title = element_text(size = 18))
#ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_psmuts.pdf", snp_plot_muts, height = 12, width = 7)
#merge
snp_plot_psgenes <- wrap_plots(snp_plot_psgenes, snp_plot_muts, ncol = 2, widths = c(0.55, 0.5))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_psgenes.pdf", snp_plot_psgenes, height = 12, width = 11)
rm(snp_plot_muts) #clean memory

# Load negative selected (NS) genes from selected pathways
snp_nsgenes <- read.csv("article_profiling/5-SNPs/dndscv/results/selected_cds_metals.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# Infer pathway from product and add dndscv attributes
snp_nspath <- tibble(gene_name = character(), pathway = character())
snp_nsstat <- tibble(wmis_cv = numeric(), pmis_cv = numeric())
for (i in 1:nrow(snp_nsgenes)) {
  #use dataframes, not lists
  snp_nspath <- snp_nspath %>%
    add_row(gene_name = snp_nsgenes$gene_name[i], pathway = fx_dndsNsPathway(snp_nsgenes$product[i]))
	snp_nsstat <- rbind(snp_nsstat, fx_dndsNsStats(snp_nsgenes$gene_name[i]))
}
snp_nsgenes <- left_join(snp_nsgenes, snp_nspath, by = c("gene_name"))
snp_nsgenes <- left_join(snp_nsgenes, snp_nsstat, by = c("gene_name"))
rm(snp_nspath, snp_nsstat) #clean memory

# Filter NS genes not called in dndscv
#skip significance (pmis_cv < 0.05)
#discard nitrogen (not enough genes)
snp_nsgenes <- snp_nsgenes %>%
  dplyr::filter(!(is.na(wmis_cv)) & wmis_cv < 1 & wmis_cv != 0 & pathway != "Nitrogen") %>%
  dplyr::mutate(product = str_replace_all(product, "%2C", ",")) #change %2C to comma

# Save NS gene ids to file
if (file.exists("article_profiling/5-SNPs/dndscv/results/kofam/ns_genes_ids.txt") == FALSE) {
  write.table(snp_nsgenes$gene_id, "article_profiling/5-SNPs/dndscv/results/kofam/ns_genes_ids.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}

# Sort pathways and set colors to plot
snp_sort_ns <- snp_nsgenes %>%
  dplyr::group_by(pathway) %>%
  dplyr::summarise(n = mean(wmis_cv)) %>%
  dplyr::arrange(desc(n)) %>%
  as.data.frame()
snp_sort_ns$color <- c("#ddad3b", "#8c2421", "#666666", "#4f4fad") #"#6a9b85"

# Violin plot for NS genes with global mutations
snp_plot_nsgenes <- ggviolin(snp_nsgenes, x = "pathway", y = "wmis_cv", color = "pathway",
                             shape = "pathway", xlab = "Pathway", ylab = "dN/dS",
                             order = snp_sort_ns$pathway, palette = snp_sort_ns$color,
                             size = 0.75, add = "jitter", legend = "none") +
  scale_y_continuous(limits = c(-0.2, 1.5)) +
  geom_hline(yintercept = 1, linetype = 1, linewidth = 0.4) +
  geom_hline(yintercept = mean(snp_nsgenes$wmis_cv), linetype = 2, linewidth = 0.4) +
  stat_compare_means(method = "anova", label.y = 1.5, size = 6) +
  stat_compare_means(comparisons = list(c("Copper", "Iron"), c("Sulfur", "Nickel"),
                                        c("Copper", "Nickel")), label = "p.format",
                     method = "t.test", size = 5) +
  theme(axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_nsgenes.pdf", snp_plot_nsgenes, height = 10, width = 7)

# Merge NS genes to unique selected gene products
snp_nsgenes_sub <- snp_nsgenes %>%
  dplyr::group_by(product, pathway) %>%
  dplyr::summarise(dnds = mean(wmis_cv))
snp_nsgenes_sub_col <- c("#8c2421", "#666666", "#4f4fad", "#ddad3b")

# Lollipop plot of grouped NS gene products
snp_plot_nsgensub <- ggdotchart(snp_nsgenes_sub, x = "product", y = "dnds", color = "pathway",
                                xlab = "Product of genes under NS", ylab = "Mean dN/dS",
                                y.text.col = TRUE, rotate = TRUE, palette = snp_nsgenes_sub_col,
                                dot.size = 3, sorting = "descending", ggtheme = theme_pubr()) +
  scale_y_continuous(limits = c(0, 1)) +
  geom_hline(yintercept = mean(snp_nsgenes$wmis_cv), linetype = 2, linewidth = 0.4) +
  geom_hline(yintercept = 1, linetype = 1, linewidth = 0.4) +
  font("legend.title", size = 14) +
  theme(legend.text = element_text(size = 14),
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_nsgenesprod.pdf", snp_plot_nsgensub, height = 11, width = 13)

# Global gene estimation of dN/dS (PS and NS)
snp_estim <- snp_selcv %>%
  dplyr::select(gene_name, wmis_cv, pmis_cv) %>%
  dplyr::filter(wmis_cv != 0)

# Populate global dataset (~2 min)
snp_gene2contig <- vector()
snp_contig2bin <- vector()
snp_genenames <- vector()
for (i in 1:nrow(snp_estim)) {
	snp_gene2contig <- append(snp_gene2contig, fx_dndsGene2Contig(snp_estim$gene_name[i]))
	snp_contig2bin <- append(snp_contig2bin, fx_dndsContig2Bin(snp_gene2contig[i]))
	snp_genenames <- append(snp_genenames, fx_dndsGeneName(snp_estim$gene_name[i]))
}
snp_estim["contig"] <- snp_gene2contig
snp_estim["bin"] <- snp_contig2bin
snp_estim["name"] <- snp_genenames
rm(snp_refcds, snp_bins, snp_genids, snp_gene2contig, snp_contig2bin, snp_genenames) #clean memory

# Add taxa & QC
snp_estim <- left_join(snp_estim, snp_dnds, by = c("bin")) %>%
  dplyr::select(-c("num", "genus", "species"))

# Normalize dN/dS per gene by MAG contamination, then group by phylum
snp_estim_phy <- snp_estim %>%
  dplyr::mutate(dnds_adj = wmis_cv * (1 - contamination / 100)) %>%
  dplyr::group_by(phylum, bin) %>%
  dplyr::summarise(snps = n(), dnds_sum = sum(dnds_adj)) %>%
  ungroup() %>%
  dplyr::mutate(dnds_glob = dnds_sum / snps) %>% #divide by total genes and get top 3 phyla
  dplyr::filter(phylum == "Proteobacteria" | phylum == "Acidobacteriota" | phylum == "Actinobacteriota")

# Sort phylum by median and set colors
snp_estim_sort_phy <- snp_estim_phy %>%
  dplyr::group_by(phylum) %>%
  dplyr::summarise(n = median(dnds_glob)) %>%
  dplyr::arrange(desc(n))
snp_estim_sort_phy$color <- c("#b15928", "#98d277", "#a6cee3")

# Boxplot global dN/dS ratios by phylum
snp_plot_estim_phy <- ggboxplot(snp_estim_phy, x = "phylum", y = "dnds_glob", color = "phylum",
                                order = snp_estim_sort_phy$phylum, palette = snp_estim_sort_phy$color,
                                add = "jitter", size = 0.75, legend = "none") +
  geom_hline(yintercept = mean(snp_estim_phy$dnds_glob), linetype = 2, linewidth = 0.4) +
  geom_hline(yintercept = 1, linetype = 1, linewidth = 0.4) +
  labs(x = "Phylum", y = "Global normalized dN/dS ratio") +
  stat_compare_means(method = "anova", label.y = 1.6, size = 6) +
  stat_compare_means(comparisons = list(c("Proteobacteria", "Actinobacteriota"),
                                        c("Proteobacteria", "Acidobacteriota"),
                                        c("Actinobacteriota", "Acidobacteriota")),
                     label = "p.format", method = "t.test", size = 5) +
  theme(axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_estim_phy.pdf", snp_plot_estim_phy, height = 10, width = 7)

# Aggregate normalized dN/dS ratios by bin
snp_estim_bin <- snp_estim %>%
  dplyr::mutate(dnds_adj = wmis_cv * (1 - contamination / 100)) %>%
  dplyr::group_by(phylum, bin) %>%
  dplyr::summarise(snps = n(), dnds_sum = sum(dnds_adj)) %>%
  ungroup() %>%
  dplyr::mutate(dnds_glob = dnds_sum / snps) %>% #divide by total genes
  dplyr::mutate(bin = str_replace_all(bin, "SPAdes-MaxBin2Refined-", "")) %>% #trim bin names
  dplyr::mutate(bin = str_replace_all(bin, "SPAdes-MetaBAT2Refined-", ""))

# Sort bins
snp_estim_sort_bin <- snp_estim_bin %>% dplyr::arrange(dnds_glob)

# Barplot global dN/dS ratios by MAG bins
snp_plot_estim_bin <- ggbarplot(snp_estim_bin, x = "bin", y = "dnds_glob", color = "phylum",
                                order = snp_estim_sort_bin$bin, palette = tree_colors,
                                fill = "phylum", rotate = TRUE) +
  geom_hline(yintercept = 1, linetype = 1, linewidth = 0.4) +
  labs(x = "MAG bin", y = "Global normalized dN/dS ratio") +
  font("legend.title", size = 14) +
  theme(legend.text = element_text(size = 14),
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_estim_bin.pdf", snp_plot_estim_bin, height = 11, width = 8)

# Load lists of PS and NS genes (externally* matched against KEGG KOs from KOfamScan hits)
snp_kops <- read.csv("article_profiling/5-SNPs/dndscv/results/kofam/ps_genes_kos.txt", sep = "\t", header = FALSE, stringsAsFactors = FALSE)
snp_kons <- read.csv("article_profiling/5-SNPs/dndscv/results/kofam/ns_genes_kos.txt", sep = "\t", header = FALSE, stringsAsFactors = FALSE)

# Reduce to unique KOs, saving scores
snp_kops <- snp_kops %>%
  dplyr::group_by(V3, V7) %>%
  dplyr::summarise(sum_ko = n(), score_avg = mean(V4), .groups = "drop") %>%
  dplyr::rename(KO = V3, KO_definition = V7) %>%
  dplyr::mutate(KO_def = paste0("(", KO, ") - ", KO_definition)) %>%
  dplyr::mutate(dNdS = "PS")
snp_kons <- snp_kons %>%
  dplyr::group_by(V3, V7) %>%
  dplyr::summarise(sum_ko = n(), score_avg = mean(V4), .groups = "drop") %>%
  dplyr::rename(KO = V3, KO_definition = V7) %>%
  dplyr::mutate(KO_def = paste0("(", KO, ") - ", KO_definition)) %>%
  dplyr::mutate(dNdS = "NS")

# Merge dataframe and filter by sum
snp_kos <- bind_rows(snp_kops, snp_kons)
snp_kos <- snp_kos %>% dplyr::filter(sum_ko > 10)
rm(snp_kops, snp_kons) #clean memory

# Histogram for KO categories by dN/dS
snp_plot_kos <- ggbarplot(snp_kos, x = "KO_def", y = "sum_ko", fill = "dNdS",
                          color = "white", palette = "aaas", sort.val = "asc",
                          rotate = TRUE, xlab = "KO definition", ylab = "Aggregated KOs") +
  font("legend.title", size = 14) +
  theme(legend.text = element_text(size = 14),
        axis.text = element_text(size = 16),
        axis.title = element_text(size = 18))
ggsave("article_profiling/5-SNPs/dndscv/results/plot/plot_dnds_kos.pdf", snp_plot_kos, height = 10, width = 15)

# Circos: filter the genome bin with global dN/dS >1
snp_cirmag <- snp_estim %>% dplyr::filter(grepl("S15.49", bin))

# Parse contigs by node id and sort dataframe
snp_nodeid <- vector()
for (i in 1:nrow(snp_cirmag)) {
  snp_nodeid <- append(snp_nodeid, str_split(snp_cirmag$contig, "_")[[i]][2])
}
snp_cirmag["node_id"] <- snp_nodeid
snp_cirmag$node_id <- as.integer(snp_cirmag$node_id)
snp_cirmag <- snp_cirmag %>% dplyr::arrange(node_id)
rm(snp_nodeid) #clean memory

# Set genomic attributes
snp_cirmagname <- "SPAdes-MetaBAT2Refined-S15.49"
snp_cirsize <- c(0, 2703949)
snp_cirran <- c(0, 0.5, 1, 1.5, 2, 2.5, 2.7) * 10^6

# Export gene names & contigs
if (file.exists("article_profiling/5-SNPs/dndscv/results/circos/S15.49_genes.txt") == FALSE) {
  write.table(snp_cirmag$gene_name, "article_profiling/5-SNPs/dndscv/results/circos/S15.49_genes.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}
if (file.exists("article_profiling/5-SNPs/dndscv/results/circos/contigs/circos_S15.49.txt") == FALSE) {
  write.table(snp_cirmag$contig, "article_profiling/5-SNPs/dndscv/results/circos/contigs/circos_S15.49.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}

# Read parsed gene annotation (externally* processed)
snp_cirgenes <- read.csv("article_profiling/5-SNPs/dndscv/results/circos/S15.49_genes_pos.txt", sep = "\t", stringsAsFactors = FALSE)

# Merge with genome selection <chr start end dnds gene>
snp_cirgenes <- snp_cirgenes %>% dplyr::mutate(chr = snp_cirmagname)
snp_cirgenes <- left_join(snp_cirgenes, snp_cirmag, by = c("gene_name"))
snp_cirgenes <- snp_cirgenes %>%
  dplyr::select(-c("pmis_cv", "bin", "phylum", "completeness", "contamination", "predicted_genes"))

# Use global positions for genes (externally* processed)
snp_cirgenpos <- read.csv("article_profiling/5-SNPs/dndscv/results/circos/contigs/circos_S15.49_relpos.txt", sep = "\t", stringsAsFactors = FALSE)
snp_cirgenes <- left_join(snp_cirgenes, snp_cirgenpos, by = c("contig"))
snp_cirgenes <- snp_cirgenes %>% dplyr::mutate(start = start + rel_pos, end = end + rel_pos)
rm(snp_cirgenpos) #clean memory

# Circos plot
circos.clear()
pdf("article_profiling/5-SNPs/dndscv/results/plot/snp_dnds_circos.pdf", height = 4, width = 4.5)
circos.par(cell.padding = c(0, 0))
circos.initialize(sectors = snp_cirmagname, xlim = snp_cirsize)
circos.track(snp_cirmagname, ylim = snp_cirsize, panel.fun = function(x, y) {
  #circos.text(snp_cirsize[1], snp_cirsize[2], cex = 0.5, adj = c(-0.05, -1.5),
  #            snp_cirmagname, facing = "bending.inside", niceFacing = TRUE)
  text(0, 0, "S15.49", cex = 0.7)
  text(0, -0.15, paste("Genome size 2703949 bp", "phylum Actinobacteriota | genus Acidithrix",
                       "168 contigs | N50 26686 | GC 47.7", sep = "\n"), cex = 0.6)
  circos.axis(h = "top", major.at = snp_cirran, labels = paste0(round(snp_cirran/10^6, 1)), #, " Mb"
              labels.cex = 0.5, lwd = 0.7, labels.facing = "clockwise") #0.35
}, bg.col = "#6a9b85", bg.border = FALSE, track.height = 0.06)
circos.genomicLabels(snp_cirgenes, labels.column = 7, cex = 0.45, line_lwd = 0.5,
                     side = "inside", line_col = "black", col = "black",
                     connection_height = 0.06, labels_height = 0.06)
dev.off()

### Enrichment ###
## PS genes for enrichment (we need KEGG KOs from KOfamScan)
#snp_psenrich <- read.csv("article_profiling/5-SNPs/dndscv/results/kofam/ps_genes_kos.txt", sep = "\t", header = #FALSE, stringsAsFactors = FALSE)
#
## Reduce to unique KOs, saving scores (converted to foldchange)
#snp_psenrich <- snp_psenrich %>%
#  dplyr::filter(V1 == "*") %>%
#  dplyr::group_by(V3, V7) %>%
#  dplyr::summarise(sum = n(), score = mean(log2(V4)), .groups = "drop") %>%
#  dplyr::rename(KO = V3, KO_definition = V7)
##create a named vector
##snp_kegglist <- snp_psenrich$score
##names(snp_kegglist) <- snp_psenrich$KO
#
## Get KEGG KO & pathway data
#snp_keggdata <- download_KEGG(species = "ko")
#snp_keggpat <- snp_keggdata$KEGGPATHID2NAME
#snp_keggkos <- snp_keggdata$KEGGPATHID2EXTID
#rm(snp_keggdata) #clean
#
## Enrich the genes by KO
##https://www.biostars.org/p/494567/
#snp_keggenrich <- enricher(gene = snp_psenrich$KO, TERM2GENE = snp_keggkos,
#                           TERM2NAME = snp_keggpat, pAdjustMethod = "BH", pvalueCutoff = 0.05)
#
## Dotplot
#pdf("article_profiling/5-SNPs/dndscv/results/plot/snp_dnds_enrichdot.pdf", height = 9, width = 7)
#dotplot(snp_keggenrich, showCategory = 30)
#dev.off()
#
## Barplot
#pdf("article_profiling/5-SNPs/dndscv/results/plot/snp_dnds_enrichbar.pdf", height = 9, width = 7)
#barplot(snp_keggenrich, showCategory = 30)
#dev.off()
#
## Category netplot
#pdf("article_profiling/5-SNPs/dndscv/results/plot/snp_dnds_enrichnet.pdf", height = 9, width = 13)
#cnetplot(snp_keggenrich, showCategory = 30, categorySize = "pvalue", foldChange = snp_psenrich$KO)
#dev.off()
#
## Upset plot (overlaps among different genes)
#upsetplot(snp_keggenrich)
```

## Figure 3 & supplementary 3

```{r fig3}
# Merge plots using patchwork
fig_3a <- wrap_plots(snp_plot_estim_bin, snp_plot_estim_phy, snp_plot_nsgenes,
                     design = "AABBCC", heights = c(1, 0.9, 0.9)) +
  plot_annotation(tag_levels = list(c("a", "b", "c"))) &
  theme(plot.tag = element_text(size = 20, face = "bold"))
ggsave("article_profiling/fig_3abc.pdf", fig_3a, height = 12, width = 21)
#ggsave("article_profiling/fig_3abc.png", fig_3a, height = 12, width = 21)
fig_3b <- wrap_plots(snp_plot_nsgensub, snp_plot_psgenes,
                     design = "DDEE", heights = c(0.9, 1)) +
  plot_annotation(tag_levels = list(c("d", "e"))) &
  theme(plot.tag = element_text(size = 20, face = "bold"))
ggsave("article_profiling/fig_3de.pdf", fig_3b, height = 12, width = 20)
#ggsave("article_profiling/fig_3de.png", fig_3b, height = 12, width = 20)
fig_3c <- wrap_plots(snp_plot_kos, design = "FFF") +
  plot_annotation(tag_levels = list(c("f"))) &
  theme(plot.tag = element_text(size = 20, face = "bold"))
ggsave("article_profiling/fig_3f.pdf", fig_3c, height = 10, width = 15)
#ggsave("article_profiling/fig_3f.png", fig_3c, height = 10, width = 15)

#sup 3
fig_s3 <- wrap_plots(snp_plot_psbins, widths = c(1)) +
  plot_annotation(tag_levels = list(c("a", "", ""))) &
  theme(plot.tag = element_text(size = 22, face = "bold"))
#ggsave("article_profiling/fig_s3.pdf", fig_s3, height = 11, width = 17)
#ggsave("article_profiling/fig_s3.png", fig_s3, height = 11, width = 17)
```
